### DISCUSSION TITLE
Realistic effects on the interaction between drone and the wind

### DISCUSSION THREAD

**Discussion by myarmatov** [microsoft/AirSim#3547] [Support Q&A]
Hi everyone. 

I want to implement a wind model on AirSim with realistic **effects** on the interaction between the drone and the wind. First of all, when the drone is moving, it should tilt at some angles according to the direction it moves. Then, when there is a wind affecting the drone, its effect should vary with the drone's tilting angle. Because, the wind is indeed the movement of air particles that creates air pressure on the drone. The force of the wind field acting on the drone then will be proportional to the cross-sectional area of the drone (**F = P*A**) which changes when the tilting angle changes. However, I could not see any of these effects on AirSim simulation.

Has anyone ever experienced this kind of issue before, could anybody suggest me a solution? Is there any build-in model/function that applies the afore-mentioned effects during the drone simulation?

Thanks in advance!

**Comment by ahmed-elsaharti** [comment]
#2867 adds wind simulation to AirSim. but AFAIK, AirSim does not consider the tilted cross-section of the drone when calculating drag in general. From what I've seen the drone is considered as a block with disk propellers as far as drag forces are concerned. I could be wrong though.

The behavior of the current wind model could be tested through the [API](https://microsoft.github.io/AirSim/apis/#wind-api)s (as seen in [set_wind.py](https://github.com/microsoft/AirSim/blob/f423d9f5a4a0d0f4fb9a2e4ddb75251040e88f81/PythonClient/multirotor/set_wind.py)) or through [settings.json](https://microsoft.github.io/AirSim/settings/#wind-settings)

**Reply by myarmatov** [reply]
Yes, I have tested the current wind model, and I also saw that the drone simulation is not realistic physically. In order to move forward, drone propellers should push the air in the opposite direction. For that, drone needs to be tilted accordingly. Since drone is not tilting during the movement, I cannot test if the wind model considers that force.

I want to implement my own low-level model to make the simulation more realistic using AirSim's some functions. If you have ever experienced implementing low-level model on AirSim, could you please suggest me "how to start"? Is there any tutorial, documentation, or etc. you can recommend?

**Reply by zimmy87** [reply]
I think a good place to start is taking a look at FastPhysicsEngine::getDragWrench(), which is where the current effects of the wind are calculated. Sounds like you want add a vector to Wrench.torque in the return result that is calculated based upon the cross-sectional area of the drone. The cross-sectional area should be able to be derived from the body parameter's drag vertices  if you understand how they are being calculated. Look at MultiRotorPhysicsBody::createDragVertices(), there is a comment there explaining how the drone is modeled as a box, and I think you should be able to use the calculate what you want.

I would also take a look at how the drag_factor_unit is being consumed in MultiRotorPhysicsBody::createDragVertices(). This variable is based on the area of the drone's box exposed to each axis along with the area of each propeller. This may be already doing part of what you want.
